tests/Workflow/MoveJatsAssetsTest.php

```diff
@@ -0,0 +1,333 @@
+
+        $document = FluentDOM::load(
+            <<<XML
+<item xmlns="http://libero.pub" xml:base="http://www.example.com/">
+    <article xmlns="http://jats.nlm.nih.gov" xmlns:xlink="http://www.w3.org/1999/xlink">
+        <body>
+            <graphic xlink:href="assets/figure1.jpg"/>
+        </body>
+    </article>
+</item>
+XML
+        );
+
+        $marking = new Marking();
+        $task = new PutTask('service', ItemId::fromString('id'), ItemVersionNumber::fromInt(1), $document);
+        $transition = new Transition('transition', 'place1', 'place2');
+
+        $event = new Event($task, $marking, $transition);
+
+        $this->mock->save(
+            new Request('GET', 'http://www.example.com/assets/figure1.jpg'),
+            new Response(200, ['Content-Type' => 'image/jpeg;foo=bar'], 'figure1')
+        );
+
+        $mover->onManipulate($event);
+
+        $path = 'id/v1/879f77a11b0649cb8af511fa5d6e4a7e.jpeg';
+
+        $this->assertSame('figure1', $this->filesystem->read($path));
+        $this->assertSame(AdapterInterface::VISIBILITY_PUBLIC, $this->filesystem->getVisibility($path));
+        $this->assertSame('image/jpeg', $this->filesystem->getMimetype($path));
+    }
+
+    /**
+     * @test
+     */
+    public function it_checks_the_origin_uri() : void
+    {
+        $mover = new MoveJatsAssets('~^http://www\.example\.com/assets/~', $this->filesystem, $this->client);
+
+        $document = FluentDOM::load(
+            <<<XML
+<item xmlns="http://libero.pub">
+    <article xmlns="http://jats.nlm.nih.gov" xmlns:xlink="http://www.w3.org/1999/xlink">
+        <body>
+            <graphic xlink:href="http://www.example.com/assets/figure.jpg"/>
+            <graphic xlink:href="http://not.example.com/assets/figure.jpg"/>
+            <graphic xlink:href="http://www.example.com/path/assets/figure.jpg"/>
+            <graphic xlink:href="http://www.example.com/path/../assets/figure.jpg"/>
+            <graphic xlink:href="assets/figure.jpg"/>
+            <graphic xlink:href="/assets/figure.jpg"/>
+        </body>
+    </article>
+</item>
+XML
+        );
+
+        $marking = new Marking();
+        $task = new PutTask('service', ItemId::fromString('id'), ItemVersionNumber::fromInt(1), $document);
+        $transition = new Transition('transition', 'place1', 'place2');
+
+        $event = new Event($task, $marking, $transition);
+
+        $this->mock->save(
+            new Request('GET', 'http://www.example.com/assets/figure.jpg'),
+            new Response(200, ['Content-Type' => 'image/jpeg'], 'figure')
+        );
+
+        $mover->onManipulate($event);
+
+        $expected = FluentDOM::load(
+            <<<XML
+<item xmlns="http://libero.pub">
+    <article xmlns="http://jats.nlm.nih.gov" xmlns:xlink="http://www.w3.org/1999/xlink">
+        <body>
+            <graphic xlink:href="http://assets/path/id/v1/cb071d80d1a54f21c8867a038f6a6c66.jpeg"
```

Comment:
these all seem to change in mysterious ways, sometimes `example.com` is substituted, sometimes it's fine. The `assets` hostname should be something that is accessible from where (browser, another service, another container)?

Reply:
Had a go at making it clearer in 1c23edacdbaabacdf957c090ed42ca7bade20c59.